### PR TITLE
[GLUTEN-3113] Check whether enabled gluten when init GlutenExecutorPlugin

### LINKED ISSUE #3113: When set `spark.plugins` with gluten, but disabled `spark.gluten.enabled`, cause some issue
### Backend

VL (Velox)

### Bug description

[Expected behavior] and [actual behavior].
In case, we set spark config as the following:
```properties
spark.plugins=io.glutenproject.GlutenPlugin
spark.gluten.enabled=false
spark.memory.offHeap.enabled=false
```
After the application start, the executor cannot start because of the following exception:
```log
java.lang.IllegalArgumentException: Must set 'spark.memory.offHeap.enabled' to true and set off heap memory size by option 'spark.memory.offHeap.size'
	at io.glutenproject.GlutenExecutorPlugin.init(GlutenPlugin.scala:195)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.$anonfun$executorPlugins$1(PluginContainer.scala:125)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.<init>(PluginContainer.scala:113)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:211)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:199)
	at org.apache.spark.executor.Executor.$anonfun$plugins$1(Executor.scala:279)
	at org.apache.spark.util.Utils$.withContextClassLoader(Utils.scala:233)
	at org.apache.spark.executor.Executor.<init>(Executor.scala:279)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$$anonfun$receive$1.applyOrElse(CoarseGrainedExecutorBackend.scala:174)
	at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:115)
	at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

At all, we may check this `spark.gluten.enabled` when init `GlutenExecutorPlugin`

### Spark version

Spark-3.3.x

### Spark configurations

spark.plugins=io.glutenproject.GlutenPlugin
spark.gluten.enabled=false
spark.memory.offHeap.enabled=false

### System information

_No response_

### Relevant logs

```bash
java.lang.IllegalArgumentException: Must set 'spark.memory.offHeap.enabled' to true and set off heap memory size by option 'spark.memory.offHeap.size'
	at io.glutenproject.GlutenExecutorPlugin.init(GlutenPlugin.scala:195)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.$anonfun$executorPlugins$1(PluginContainer.scala:125)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.apache.spark.internal.plugin.ExecutorPluginContainer.<init>(PluginContainer.scala:113)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:211)
	at org.apache.spark.internal.plugin.PluginContainer$.apply(PluginContainer.scala:199)
	at org.apache.spark.executor.Executor.$anonfun$plugins$1(Executor.scala:279)
	at org.apache.spark.util.Utils$.withContextClassLoader(Utils.scala:233)
	at org.apache.spark.executor.Executor.<init>(Executor.scala:279)
	at org.apache.spark.executor.CoarseGrainedExecutorBackend$$anonfun$receive$1.applyOrElse(CoarseGrainedExecutorBackend.scala:174)
	at org.apache.spark.rpc.netty.Inbox.$anonfun$process$1(Inbox.scala:115)
	at org.apache.spark.rpc.netty.Inbox.safelyCall(Inbox.scala:213)
	at org.apache.spark.rpc.netty.Inbox.process(Inbox.scala:100)
	at org.apache.spark.rpc.netty.MessageLoop.org$apache$spark$rpc$netty$MessageLoop$$receiveLoop(MessageLoop.scala:75)
	at org.apache.spark.rpc.netty.MessageLoop$$anon$1.run(MessageLoop.scala:41)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

### PR DESCRIPTION
**Pull request by zwangsheng**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#ISSUE-ID)
close #3113

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
~No, be happy to add unit tests if needed~
Add unit test.

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by zwangsheng** [comment]
The user may think that setting `spark.gluten.enabled` to false turned off gluten completely, and before this PR, the user had to manually remove gluten plugin from `spark.plugins` list to turned off gluten completely.

File: gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala [hunk 1]
```diff
@@ -57,11 +57,18 @@ private[glutenproject] class GlutenDriverPlugin extends DriverPlugin with Loggin
   private var _sc: Option[SparkContext] = None
 
   override def init(sc: SparkContext, pluginContext: PluginContext): util.Map[String, String] = {
+    val conf = pluginContext.conf()
+    if (!conf.getBoolean(GLUTEN_ENABLE_KEY, defaultValue = true)) {
+      logWarning(
+        "Configured to not enabled Gluten, but gluten added in `spark.plugins`! " +
```
Comment (winningsix) [inline gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala:63]: Nit: ```Configured not to enable Gluten plugin in driver``` better ?

File: gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala [hunk 2]
```diff
@@ -176,14 +183,19 @@ private[glutenproject] class GlutenDriverPlugin extends DriverPlugin with Loggin
...
   private var executorEndpoint: GlutenExecutorEndpoint = _
   private val taskListeners: Seq[TaskListener] = Array(TaskResources)
 
   /** Initialize the executor plugin. */
   override def init(ctx: PluginContext, extraConf: util.Map[String, String]): Unit = {
     val conf = ctx.conf()
-
+    if (!conf.getBoolean(GLUTEN_ENABLE_KEY, defaultValue = true)) {
+      logWarning(
+        "Configured to not enabled Gluten, but gluten added in `spark.plugins`! " +
```
Comment (winningsix) [inline gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala:195]: ditto. but specifically add executor in log.

Comment (zwangsheng) [reply gluten-core/src/main/scala/io/glutenproject/GlutenPlugin.scala]: Thanks for review, address those.

**Comment by zwangsheng** [comment]
Added unit test, PTAL @winningsix @ulysses-you

**Comment by ulysses-you** [comment]
`spark.gluten.enabled` it is a sql config that means it can be changed at runtime. If we want to make it as a global switch, we should mark it as static config. Use `buildStaticConf` to replace `buildConf`. I'm not sure if the change will break anything. also cc @zhouyuan @PHILO-HE

**Comment by PHILO-HE** [comment]
Do you have any reason to use this config in practice? I am thinking whether we can remove this config and just use `spark.plugins` to control the usage of gluten plugin. Thanks!

**Comment by ulysses-you** [comment]
yeah, use `spark.plugins` seems a good way.

**Comment by zwangsheng** [comment]
> Do you have any reason to use this config in practice? I am thinking whether we can remove this config and just use spark.plugins to control the usage of gluten plugin. Thanks!

@PHILO-HE thanks for comment, remove `spark.gluten.enabled` sounds good to me, in consideration of this may make users have the wrong perception.

**Comment by zwangsheng** [comment]
Close this PR, ref to the above discussion.

**Comment by PHILO-HE** [comment]
As discussed with @zhztheplayer, another thought is just keeping this `spark.gluten.enabled` config at source code. It can be used to dynamically control Gluten's usage at a spark context. This config should have NO impact on Gluten plugin's registry and initialization, etc., but just expect to control something introduced by Gluten, such as plan replacement, join selection strategy, etc. It indeed has some functionality gap.
If no strong reason, let's just use `spark.plugins` instead to control the whole usage. Thanks!

**Comment by zwangsheng** [comment]
May be we can update/improve `spark.gluten.enabled` doc to inform user this can turn off gluten plugin.XD

**Comment by PHILO-HE** [comment]
> May be we can update/improve `spark.gluten.enabled` doc to inform user this can't turn off gluten plugin. XD

Yes, I will update the docs. Thanks for your feedback!

**Pull request closed without merging**